### PR TITLE
maddy: 0.6.3 -> 0.7.0

### PR DESCRIPTION
**Pull request by NickCao**
###### Description of changes
contains a modified version of https://github.com/NixOS/nixpkgs/pull/234439

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by onny** [comment]
Result of `nixpkgs-review pr 234930` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>maddy</li>
  </ul>
</details>

File: nixos/modules/services/mail/maddy.nix [hunk 1]
```diff
@@ -335,12 +335,13 @@ in {
       };
 
       secrets = lib.mkOption {
-        type = lib.types.path;
+        type = with types; listOf path;
```
Comment (e1mo) [inline nixos/modules/services/mail/maddy.nix:338]: Thinking about it, that way we break backwards compatibility with setups that use the old style of config declaration (blocking(?) backportability).
That way we can keep compatibility:
```suggestion
        type = with types; (either (nullOr path) (listOf path)) // {
          description = "list of path";
        };
        apply = x: if isList x then x else singleton (warn "Passing a single path to `services.maddy.secrets' is deprecated. Please pass a list of paths instead." x);
```

**Review by e1mo — COMMENTED** [review]
Package builds and tests are passing, just one remark regarding the options change.

File: nixos/modules/services/mail/maddy.nix (hunk 1 above)
Comment (NickCao) [reply nixos/modules/services/mail/maddy.nix:338]: This feels too complicated, or we only support listing a single file as it was?

Comment (e1mo) [reply nixos/modules/services/mail/maddy.nix:338]: Then I'd prefer going with the list and just let the error message speak for itself.

**Review by e1mo — APPROVED** [review]

**Pull request merged**